opm/autodiff/AquiferCarterTracy.hpp

```diff
@@ -0,0 +1,505 @@
+/*
```

Comment:
I think it is good to separate the implementation to a  `_impl.hpp`  to make this class interface easier to read.

Reply:
> I think it is good to separate the implementation to a  `_impl.hpp`  to make this class interface easier to read.

I disagree because `.hpp`/`_impl.hpp`/`.cpp` splits have frequently made my life *very* miserable for stuff that I needed to modify. That said, I will not complain this (anti-) pattern for code that I'm not  supposed to touch often, but could we please agree that whether to have all function definitions in separate files or not is in the discretion of the main author of the code?

Reply:
> I disagree because .hpp/_impl.hpp/.cpp splits have frequently made my life very miserable for stuff that I needed to modify.

If you separate the definition and implementation in different files in complete way (I mean not mixed, part of implementation in `.hpp` file, part of the implementation in `_impl.hpp` file), I do not understand why you were very miserable.  Could you please explain this?   There will be some extra manual work, but comparing to the cost of learning the design/interface of a new class with 200 lines and a new class with 2000 lines, I think the extra work is small.  I know you suggested some software tools or Doxygen, but not everyone is using them and they also come with extra efforts. It is good let the code speak for themselves, like when we read a PR, what tools should we use?

Reply:
I agree with @andlaus . I had a very bad time tracking the #defines when I was starting out with the code.

Reply:
keep .cpp out of the discussion. splitting into sane compile units is a *tool*, not a preference.

Reply:
I think @kel85uk summarizes my troubles nicely and I do not consider myself to be a noob in this code base.

> keep .cpp out of the discussion. splitting into sane compile units is a tool, not a preference.

`.cpp` files are related to `_impl.hpp` files because if you change a non-template class to a template one you need to convert the former into the latter and this happened quite frequently in the past. another issue with this pattern are non-template classes that have some templated methods. (the latter cannot be moved to cpp, agree?)

the reason I really dislike the "separate files for the implementation" pattern is that -- besides making things  much more complicated than necessary -- it adds a lot of inertia (try to change the template parameters of such classes) and potential for error (e.g. the ordering of the template arguments can be reversed in the `_impl.hpp` file, this happened e.g. in the EQUIL init code before it was moved to eWoms) for no obvious benefit.

Reply:
i said nothing about _impl files.

what i dislike is rebuilding everything and its dog just because i change a small detail. that's far far more annoying than having to relate to two editors, and is a true anti-pattern, a pattern you are advocating by including cpp files in that statement. they are, as a i said, a tool. you use it when you can, if not you do the other thing.

Reply:
> what i dislike is rebuilding everything and its dog just because i change a small detail. that's far far more annoying than having to relate to two editors, and is a true anti-pattern, a pattern you are advocating by including cpp files in that statement. they are, as a i said, a tool. you use it when you can, if not you do the other thing.

This probably boils down to cultural differences: Only a single compile unit is involved for a target "everything and its dog" is still just a single compile unit. Probably we currently suffer from an "in between" approach.

It seems though that we both agree to leave the decision of how to deal with organizing code to the author of that code?

Reply:
i don't agree it's a cultural thing. it's a technical thing. compile units is the chunks you can parallelize on. i like to use all my cores for building, and unless there are reasons (ie templates), one should strive for many small compile units because then we get to use our fancy multi-core machines most efficiently. everybody will have some deal with some in-between as what is done is dictated by the language and its limitations (iability to inline and such falls in this category), not personal choice.

Reply:
> i don't agree it's a cultural thing. it's a technical thing.

alright, technical. you asked for it ;)

> compile units is the chunks you can parallelize on. 

I agree that this is true for C, but in C++ this dogma break down for template code (i.e., "almost always" in the case of OPM) and even if you don't use templates, the compiler being able to inline everything can have a quite large performance impact. @atgeirr learned this lesson recently, IIRC; of course most of this can amended using LTO as a band-aid, but LTO is essentially the same as using a single compile unit from the outset and prone to cause troubles in my experience.

> i like to use all my cores for building,

the same amount of parallelization can also be achieved when compiling multiple programs at once. For builds that include the tests, there are almost always enough targets.

> one should strive for many small compile units because then we get to use our fancy multi-core machines most efficiently.

I see your point, but in practice it has only limited validity: recompiling "plain ebos" is much faster when changing something in e.g. opm-material than recompilation of flow. that is due to the fact that libopmsimulators goobbles up most of the time, even when disregarding the `flow_ebos_*.cpp` CUs. If I need to recompile everything because I want to switch compilers or flags, the difference is only exacerbated.

> everybody will have some deal with some in-between as what is done is dictated by the language and its limitations (iability to inline and such falls in this category), not personal choice.

Eigen, some dune core modules and opm-material as well as eWoms do not require a library, so I think that in fact this *is* a choice that mainly depends on personal taste. Also, note that most "newer" languages like Java, Go, Rust or python do not allow a source/header split from the outset, and I think that there is a good reason for this.

Reply:
> I agree with @andlaus . I had a very bad time tracking the #defines when I was starting out with the code. The _impls wasn't very clear with the filename either for a noob such as myself...

Can you please detail the problem related to the `#defines`? Sorry I am not following this.

Reply:
> Also, note that most "newer" languages like Java, Go, Rust or python do not allow a source/header split from the outset, and I think that there is a good reason for this.  

This is not a good argument, since you probably can also argue why Java does not have `#include`, why OPM is not written in Java. 

Anyway, I am more relaxed about the `_impl.hpp` issue now, as long as the class definition layout is well made (I mean the public/private/protected member variables and member functions are well categorized in the correct location, not show up randomly). Although I still prefer a clean and shorter header file (tens of lines for a small class, a couple of hundred lines for a relatively complicated class), which makes the people learn how to use / change the new class much easier, and also it is easier to keep the class clean.  With all the features of the modern editors, I have never felt that the extra manual work to edit two files is a burden compared to other efforts involved in the development. But we can agree to disagree here. 

I agree `.cpp` file should not have been in the same discussion and that is why I did not mention it at all.

Reply:
> This is not a good argument, since you probably can also argue why Java does not have #include, why OPM is not written in Java.

OPM targets the highest performance possible and this requires a language that is basically portable assembler. the fact that C++ can be used to write higher higher performance code than most other languages that were around at the start of the project does not mean all its historical features are good ideas from the conceptual point of view (the preprocessor is one of the bad ones, for example, or you can still use gotos). I guess if the project was started from scratch now, we would take a hard look at Rust an Go...

> Anyway, I am more relaxed about the _impl.hpp issue now, as long as the class definition layout is well made. [...] Although I still prefer a clean and shorter header file [...] which makes the people learn how to use / change the new class much easier.

I'd recommend to use an editor that supports code folding and to hide the method bodies if you want to get an overview...

> With all the features of the modern editors, I have never felt that the extra manual work to edit two files is a burden compared to other efforts involved in the development.

I use emacs, and switching back and forth between files is a *major* productivity killer for me. in particular, that's the case if headers are located in completely different directories than the sources or if the implementation of multiple classes is combined into a single source file. vi users -- even though they're normally the enemy ;) -- will probably agree here. if you argue that one should use a more "modern" editor, why don't you use one that supports code folding?

Reply:
> the fact that C++ can be used to write higher higher performance code than most other languages that were around at the start of the project does not mean all its historical features are good ideas from the conceptual point of view (the preprocessor is one of the bad ones, for example, or you can still use gotos).

No language is perfect, that is why the discussion should be within the scope of the specific language. 

> I use emacs, and switching back and forth between files is a major productivity killer for me. in particular, that's the case if headers are located in completely different directories than the sources or if the implementation of multiple classes is combined into a single source file.

There is no discussion about the headers in a different folder here.  I am not aware of that emacs does not support screen splitting. It is a great feature with the HD display or display with higher resolution. 

> if you argue that one should use a more "modern" editor, 

I should not have had `modern` there. Vim is a very good to go. 

Since I agree to disagree, my argument finishes here.